### PR TITLE
feat(aggexec): add approx_percentile aggregate function for p95/p99

### LINKED ISSUE #24550: [Feature Request]: Percentile / quantile aggregate functions (p95 / p99)
**Version:** `8.0.30-MatrixOne-v3.0.11` (MatrixOne v3.0.11)

Only `median` (p50) is available; there is no general percentile/quantile function. p95/p99 are core metrics for latency/SLA analytics and observability.

**Reproduce**
```sql
SELECT median(v) FROM t;                                          -- OK (p50 only)
SELECT quantile(0.95)(v) FROM t;                                  -- SQL parser error
SELECT approx_percentile(v, 0.95) FROM t;                         -- not supported
SELECT percentile_cont(0.95) WITHIN GROUP (ORDER BY v) FROM t;    -- SQL parser error
```

**Request**
Add percentile/quantile aggregates (ideally both exact and an approximate / t-digest variant) so p95/p99 can be computed.

---
_Found while evaluating MatrixOne git4data / SQL as a data-versioning & agent-trace backend. Version: `8.0.30-MatrixOne-v3.0.11` (MatrixOne v3.0.11)._

## Testing background
Found while building **observability rollups over agent traces** (p95 / p99 latency per model / per tool) and while benchmarking MatrixOne against ClickHouse as an OTel trace backend. Only `median` (p50) is available.

## Why it matters (expected purpose)
p95/p99 latency are **core SLA / observability metrics** and standard analytics. Without percentile/quantile (ideally an approximate / t-digest variant for scale) MatrixOne can't serve these dashboards, which is a key gap versus OLAP engines used for tracing.

### PR DESCRIPTION
**Pull request by ULookup**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #24550

## What this PR does / why we need it:

Adds the approx_percentile(v, percentile) aggregate function that computes
the approximate percentile of a numeric column using O(n) quickselect
with linear interpolation. This enables p95/p99 latency analytics and
observability dashboards previously limited to median (p50) only.

The implementation wires together the parser grammar (already present),
plan-layer function registration, execution-layer executor (reusing the
median infrastructure), and compile-layer config extraction for the
percentile constant.

### CODE REVIEW

File: pkg/sql/compile/operator.go [hunk 1]
```diff
@@ -1469,14 +1476,20 @@ func constructGroup(_ context.Context, node, childNode *plan.Node, needEval bool
 			if len(f.F.Args) > 0 {
 				//for group_concat, the last arg is separator string
 				//for cluster_centers, the last arg is kmeans_args string
+				//for approx_percentile, the last arg is the percentile value
 				if (f.F.Func.ObjName == plan2.NameGroupConcat ||
-					f.F.Func.ObjName == plan2.NameClusterCenters) && len(f.F.Args) > 1 {
+					f.F.Func.ObjName == plan2.NameClusterCenters ||
+					f.F.Func.ObjName == "approx_percentile") && len(f.F.Args) > 1 {
 					argExpr := f.F.Args[len(f.F.Args)-1]
 					vec, free, err := colexec.GetReadonlyResultFromNoColumnExpression(proc, argExpr)
```
Comment (XuPeng-SH) [inline pkg/sql/compile/operator.go:1498]: This assumes the last argument is a non-null no-column constant, but the new SQL contract does not enforce that. On this head, a NULL percentile constant already panics in `getPercentileConfig` because the returned const-null vector has no fixed payload to index. And a row-dependent percentile expression is also admitted by `approx_percentile`'s type check even though `GetReadonlyResultFromNoColumnExpression` explicitly forbids column expressions. Please validate arg2 before this path instead of panicking here.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
I found one blocking contract issue in the new `approx_percentile` aggregate. The execution path assumes the percentile argument is a non-null no-column constant, but the SQL function contract does not enforce that.

- `list_agg.go` accepts `T_any` for arg2 and does not require constness.
- `constructGroup` / `constructWindow` then strip the last argument and feed it through `colexec.GetReadonlyResultFromNoColumnExpression`.
- On the current head, a minimal package-local repro still panics for a NULL percentile constant: `getPercentileConfig(vector.NewConstNull(types.T_float64.ToType(), 1, mp))` crashes at `operator.go:2332` with `index out of range [0] with length 0`.
- The same missing contract also admits row-dependent percentile expressions. `GetReadonlyResultFromNoColumnExpression` explicitly forbids column expressions; for a simple `Expr_Col` it indexes the const-fold batch directly, so `approx_percentile(v, p_col)` is on the same bad path.

Please reject NULL / non-constant percentile arguments during planning (or at least return a user error instead of panicking), and add regressions for both `approx_percentile(v, null)` and `approx_percentile(v, p_col)`.

**Review by heni02 — APPROVED** [review]

File: test/distributed/cases/function/func_aggr_approx_percentile.test [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+select approx_percentile(b, 0.5) from t1;
+select approx_percentile(b, 0.0) from t1;
+select approx_percentile(b, 1.0) from t1;
+
+insert into t1 values (1, 2), (1, 4), (2, 6), (2, 8);
+select a, approx_percentile(b, 0.5) from t1 group by a order by a;
+
+drop table if exists t1;
+create table t1 (a int, b float);
+insert into t1 values (1, 1.1), (2, 2.2), (3, 3.3);
```
Comment (fengttt) [inline test/distributed/cases/function/func_aggr_approx_percentile.test:15]: You have to add a reasonable tests, for example, 10K rows, with duplicates, with nulls, etc. and check the result it correct.

**Review by fengttt — CHANGES_REQUESTED** [review]

File: pkg/sql/colexec/aggexec/approx_percentile.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+		if vLo, err = vLo.Scale(1); err != nil {
+			return types.Decimal128{}, err
+		}
+		return vLo, nil
+	}
+	vHi := selectKthFunc(vals, hi, compare)
+	frac := idx - float64(lo)
+
+	// Use float64-based interpolation for accuracy.
+	fLo := types.Decimal128ToFloat64(vLo, argScale)
```
Comment (XuPeng-SH) [inline pkg/sql/colexec/aggexec/approx_percentile.go:143]: [blocker] Converting `Decimal128` through `float64` corrupts valid decimal inputs. Repro: values `9007199254740993.00, 9007199254740994.00`, `p=0`; expected `9007199254740993.000`, actual `9007199254740994.048`. The same precision loss affects general interpolation (and Decimal64 once its significant digits exceed float64 precision), so a p=0 special case alone would not close this. Please keep the interpolation in decimal/fixed-point arithmetic and add >2^53 regressions.

File: pkg/sql/colexec/aggexec/approx_percentile.go [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+	fHi := types.Decimal128ToFloat64(vHi, argScale)
+	result := fLo + frac*(fHi-fLo)
+	return types.Decimal128FromFloat64(result, 38, argScale+1)
+}
+
+// --- Executor types ---
+
+type approxPercentileNumericExec[T numeric] struct {
+	medianColumnExecSelf[T, float64]
```
Comment (XuPeng-SH) [inline pkg/sql/colexec/aggexec/approx_percentile.go:152]: [blocker] Reusing `medianColumnExecSelf` makes this aggregate retain every input value per group. Its inherited intermediate-state methods also serialize every value, and `collectMedianValues` allocates another complete `[]T` at finalize. Thus `approx_percentile` has O(rows) state/shuffle and roughly doubles peak local storage instead of providing a bounded mergeable approximation. This needs a bounded sketch (or an exact-percentile API/name with explicit resource semantics) before it is safe for the intended large observability rollups.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
The previous NULL/non-constant percentile contract blocker is fixed on the current head, and this branch merges cleanly with current `main`. I still found two merge blockers:

1. Decimal interpolation is incorrect for values that cannot be represented exactly as `float64`. A package-local regression with `DECIMAL128` values `9007199254740993.00` and `9007199254740994.00` at `p = 0` expects `9007199254740993.000`, but the current code returns `9007199254740994.048`. This is user-visible corruption, including at the exact min boundary.
2. The executor is not approximate or bounded: it embeds the median executor, retains every non-NULL value for every group, serializes all retained rows as intermediate state, and copies them all again during finalize. Therefore both local state and distributed shuffle are O(rows), with an additional O(rows) Go-heap allocation outside the mpool during flush. That defeats the scale motivation of `approx_percentile` and can OOM on the p95/p99 observability workloads this feature targets. Please use a bounded, mergeable sketch (for example t-digest/KLL), or expose/document this as an exact percentile API rather than `approx_percentile`.

Please also add a realistic regression (10K+ rows with duplicates and NULLs, including partial-state merge) and large `DECIMAL64/128` cases above 2^53. Existing tests pass, but the largest dataset is only 20 rows and does not exercise either blocker.

**Review by aunjgr — CHANGES_REQUESTED** [review]
This implementation retains every input row, serializes unbounded intermediate state, and converts Decimal128 interpolation through float64, corrupting values above 2^53. PR #25881 is the bounded replacement and Mergify reports that this PR blocks evaluation of it. Please close this PR in favor of #25881.

**Review by aunjgr — CHANGES_REQUESTED** [review]
This implementation retains every input row, serializes unbounded intermediate state, and converts Decimal128 interpolation through float64, corrupting values above 2^53. PR #25881 is the bounded replacement and Mergify reports that this PR blocks evaluation of it. Please close this PR in favor of #25881.
